core/embed/rust/src/translations/blob.rs

```diff
@@ -0,0 +1,381 @@
+        if magic != Self::HEADER_MAGIC {
+            return Err(INVALID_TRANSLATIONS_BLOB);
+        }
+
+        let model = read_fixedsize_str(&mut header_reader, 4)?;
+
+        if model != crate::trezorhal::model::INTERNAL_NAME {
+            return Err(value_error!("Wrong Trezor model"));
+        }
+
+        let version_bytes = header_reader.read(4)?;
+        let version = unwrap!(version_bytes.try_into());
+
+        let data_len = header_reader.read_u16_le()? as usize;
+        let data_hash: sha256::Digest =
+            unwrap!(header_reader.read(sha256::DIGEST_SIZE)?.try_into());
+
+        let language = read_pascal_str(&mut header_reader)?;
+        if language.len() > 8 {
+            return Err(INVALID_TRANSLATIONS_BLOB);
+        }
+        let change_language_title = read_pascal_str(&mut header_reader)?;
+        let change_language_prompt = read_pascal_str(&mut header_reader)?;
+
+        // ignore the rest of the header reader - this allows older firmware to
+        // understand newer header if there are only added items
+        _ = header_reader.rest();
+
+        //
+        // 3. parse the proof section
+        //
+        let mut proof_reader = read_u16_prefixed_block(&mut reader)?;
+        let proof_count = proof_reader.read_byte()? as usize;
+        let proof_length = proof_count * sha256::DIGEST_SIZE;
+        let proof_bytes = proof_reader.read(proof_length)?;
+
+        // create a list of the proof items
+        // SAFETY: sha256::Digest is a plain array of u8, so any bytes are valid
+        let (_prefix, merkle_proof, _suffix) = unsafe { proof_bytes.align_to::<sha256::Digest>() };
+        if !_prefix.is_empty() || !_suffix.is_empty() {
+            return Err(INVALID_TRANSLATIONS_BLOB);
+        }
+        let signature = cosi::Signature::new(
+            proof_reader.read_byte()?,
+            unwrap!(proof_reader.read(ed25519::SIGNATURE_SIZE)?.try_into()),
+        );
+
+        // check that there is no trailing data in the proof section
+        if proof_reader.remaining() > 0 {
+            return Err(INVALID_TRANSLATIONS_BLOB);
+        }
+
+        // check that the declared data section length matches the container size
+        if container_length - reader.tell() != data_len {
+            return Err(INVALID_TRANSLATIONS_BLOB);
+        }
+
+        let new = Self {
+            header_bytes,
+            language,
+            version,
+            data_len,
+            data_hash,
+            change_language_title,
+            change_language_prompt,
+            merkle_proof,
+            signature,
+            total_len: container_length + Self::BLOB_MAGIC.len() + mem::size_of::<u16>(),
+        };
+        new.verify()?;
+        Ok((new, reader))
+    }
+
+    fn verify_with_keys(&self, public_keys: &[ed25519::PublicKey]) -> Result<(), Error> {
+        let merkle_root = merkle_root(self.header_bytes, self.merkle_proof);
```

Comment:
Why do we even use a Merkle tree? Because producing a signature is expensive so we just create one for a batch of blobs?

Reply:
"expensive", yes, as in, this is also going to be signed manually so we don't want to overwhelm the signers